Bottom: 6f69309199c3fd37bc52eb39f837767edcb48829
Top:    4c35280631dd974679f421faa3606b0d73bddbe2
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-25 18:08:22 +0100

draw.c: Split "Gathering" routines from real drawing routines (Polygons)

Moves code from DrawPlainPolygon() into its only caller poly_callback()


---

diff --git a/src/draw.c b/src/draw.c
index 81983c2..c5b908c 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -87,8 +87,6 @@ static void DrawEverything (BoxTypePtr);
 static void DrawPPV (int group, const BoxType *);
 static int DrawLayerGroup (int, const BoxType *);
 static void DrawRegularText (LayerTypePtr, TextTypePtr);
-static void DrawPolygonLowLevel (PolygonTypePtr);
-static void DrawPlainPolygon (LayerTypePtr Layer, PolygonTypePtr Polygon);
 static void AddPart (void *);
 static void SetPVColor (PinTypePtr, int);
 static void DrawEMark (ElementTypePtr, LocationType, LocationType, bool);
@@ -697,12 +695,45 @@ clearPin_callback (const BoxType * b, void *cl)
     gui->fill_pcb_pv (Output.pmGC, Output.pmGC, pin, false, true);
   return 1;
 }
+
 static int
 poly_callback (const BoxType * b, void *cl)
 {
   LayerType *layer = cl;
+  PolygonType *polygon = (PolygonType *b);
+  static char *color;
+
+  if (!polygon->Clipped)
+    return 0;
+
+  if (TEST_FLAG (SELECTEDFLAG, Polygon))
+    color = Layer->SelectedColor;
+  else if (TEST_FLAG (FOUNDFLAG, Polygon))
+    color = PCB->ConnectedColor;
+  else
+    color = Layer->Color;
+  gui->set_color (Output.fgGC, color);
+
+  if (gui->thindraw_pcb_polygon != NULL &&
+      (TEST_FLAG (THINDRAWFLAG, PCB) ||
+       TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
+    gui->thindraw_pcb_polygon (Output.fgGC, polygon, clip_box);
+  else
+    gui->fill_pcb_polygon (Output.fgGC, polygon, clip_box);
+
+  /* If checking planes, thin-draw any pieces which have been clipped away */
+  if (gui->thindraw_pcb_polygon != NULL &&
+      TEST_FLAG (CHECKPLANESFLAG, PCB) &&
+      !TEST_FLAG (FULLPOLYFLAG, Polygon))
+    {
+      PolygonType poly = *polygon;
+
+      for (poly.Clipped = polygon->Clipped->f;
+           poly.Clipped != polygon->Clipped;
+           poly.Clipped = poly.Clipped->f)
+        gui->thindraw_pcb_polygon (Output.fgGC, &poly, clip_box);
+    }
 
-  DrawPlainPolygon (layer, (PolygonTypePtr) b);
   return 1;
 }
 
@@ -1156,16 +1187,9 @@ DrawTextLowLevel (TextTypePtr Text, int min_line_width)
 static void
 DrawPolygonLowLevel (PolygonTypePtr Polygon)
 {
-  if (!Polygon->Clipped)
-    return;
-
-  if (Gathering)
-    {
-      AddPart (Polygon);
-      return;
-    }
+  assert (Gathering);
 
-  printf ("DrawPolygonLowLevel: Called without Gathering set!\n");
+  AddPart (Polygon);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1358,7 +1382,12 @@ DrawRegularText (LayerTypePtr Layer, TextTypePtr Text)
 void
 DrawPolygon (LayerTypePtr Layer, PolygonTypePtr Polygon)
 {
-  DrawPolygonLowLevel (Polygon);
+  assert (Gathering);
+
+  if (!Polygon->Clipped)
+    return;
+
+  AddPart (Polygon);
 }
 
 int
@@ -1389,53 +1418,6 @@ thin_callback (PLINE * pl, LayerTypePtr lay, PolygonTypePtr poly)
   return 0;
 }
 
-
-/* ---------------------------------------------------------------------------
- * draws a polygon
- */
-static void
-DrawPlainPolygon (LayerTypePtr Layer, PolygonTypePtr Polygon)
-{
-  static char *color;
-
-  if (!Polygon->Clipped)
-    return;
-
-  if (Gathering)
-    {
-      AddPart (Polygon);
-      return;
-    }
-
-  if (TEST_FLAG (SELECTEDFLAG, Polygon))
-    color = Layer->SelectedColor;
-  else if (TEST_FLAG (FOUNDFLAG, Polygon))
-    color = PCB->ConnectedColor;
-  else
-    color = Layer->Color;
-  gui->set_color (Output.fgGC, color);
-
-  if (gui->thindraw_pcb_polygon != NULL &&
-      (TEST_FLAG (THINDRAWFLAG, PCB) ||
-       TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
-    gui->thindraw_pcb_polygon (Output.fgGC, Polygon, clip_box);
-  else
-    gui->fill_pcb_polygon (Output.fgGC, Polygon, clip_box);
-
-  /* If checking planes, thin-draw any pieces which have been clipped away */
-  if (gui->thindraw_pcb_polygon != NULL &&
-      TEST_FLAG (CHECKPLANESFLAG, PCB) &&
-      !TEST_FLAG (FULLPOLYFLAG, Polygon))
-    {
-      PolygonType poly = *Polygon;
-
-      for (poly.Clipped = Polygon->Clipped->f;
-           poly.Clipped != Polygon->Clipped;
-           poly.Clipped = poly.Clipped->f)
-        gui->thindraw_pcb_polygon (Output.fgGC, &poly, clip_box);
-    }
-}
-
 /* ---------------------------------------------------------------------------
  * draws an element
  */
@@ -1673,7 +1655,9 @@ EraseText (LayerTypePtr Layer, TextTypePtr Text)
 void
 ErasePolygon (PolygonTypePtr Polygon)
 {
-  DrawPolygonLowLevel (Polygon);
+  assert (Gathering);
+
+  AddPart (Polygon);
 }
 
 /* ---------------------------------------------------------------------------
